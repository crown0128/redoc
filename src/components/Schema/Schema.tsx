import * as React from 'react';
import { observer } from 'mobx-react';

import { FieldDetails } from '../Fields/FieldDetails';
import { TypeName, CircularLabel } from '../../common-elements/fields';

import { SchemaModel } from '../../services/models';

import { ObjectSchema } from './ObjectSchema';
import { OneOfSchema } from './OneOfSchema';
import { ArraySchema } from './ArraySchema';

export interface SchemaProps {
  schema: SchemaModel;
  showTitle?: boolean;
}

@observer
export class Schema extends React.Component<Partial<SchemaProps>> {
  render() {
    const { schema } = this.props;
    if (!schema) return <em> Schema not provided </em>;
    const { type, oneOf, discriminatorProp, isCircular } = schema;

    if (isCircular) {
      return (
        <div>
          <TypeName>{schema.displayType}</TypeName>
          <CircularLabel> Circular </CircularLabel>
        </div>
      );
    }

    if (discriminatorProp !== undefined) {
      return (
        <ObjectSchema
          {...{ ...this.props, schema: oneOf![schema.activeOneOf] }}
          discriminator={{
            fieldName: discriminatorProp,
            parentSchema: schema,
          }}
        />
      );
    }

    if (oneOf !== undefined) {
      return <OneOfSchema schema={schema} />;
    }

    switch (type) {
      case 'object':
        return <ObjectSchema {...this.props as any} />;
      case 'array':
        return <ArraySchema {...this.props as any} />;
    }

    // TODO: maybe adjust FieldDetails to accept schema
    return (
      <div>
        <FieldDetails
          field={{
            schema,
            name: '',
            required: false,
            description: schema.description,
            deprecated: false,
            toggle: () => null,
            expanded: false,
          }}
        />
      </div>
    );
  }
}
