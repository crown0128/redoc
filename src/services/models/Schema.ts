import { action, observable } from 'mobx';

import { OpenAPIExternalDocumentation, OpenAPISchema, Referenced } from '../../types';

import { OpenAPIParser } from '../OpenAPIParser';
import { RedocNormalizedOptions } from '../RedocNormalizedOptions';
import { FieldModel } from './Field';

import { MergedOpenAPISchema } from '../';
import {
  detectType,
  extractExtensions,
  humanizeConstraints,
  isNamedDefinition,
  isPrimitiveType,
  JsonPointer,
  sortByField,
  sortByRequired,
} from '../../utils/';

// TODO: refactor this model, maybe use getters instead of copying all the values
export class SchemaModel {
  pointer: string;

  type: string;
  displayType: string;
  typePrefix: string = '';
  title: string;
  description: string;
  externalDocs?: OpenAPIExternalDocumentation;

  isPrimitive: boolean;
  isCircular: boolean = false;

  format?: string;
  displayFormat?: string;
  nullable: boolean;
  deprecated: boolean;
  pattern?: string;
  example?: any;
  enum: any[];
  default?: any;
  readOnly: boolean;
  writeOnly: boolean;

  constraints: string[];

  fields?: FieldModel[];
  items?: SchemaModel;

  oneOf?: SchemaModel[];
  oneOfType: string;
  discriminatorProp: string;
  @observable
  activeOneOf: number = 0;

  rawSchema: OpenAPISchema;
  schema: MergedOpenAPISchema;
  extensions?: Dict<any>;

  /**
   * @param isChild if schema discriminator Child
   * When true forces dereferencing in allOfs even if circular
   */
  constructor(
    parser: OpenAPIParser,
    schemaOrRef: Referenced<OpenAPISchema>,
    pointer: string,
    private options: RedocNormalizedOptions,
    isChild: boolean = false,
  ) {
    this.pointer = schemaOrRef.$ref || pointer || '';
    this.rawSchema = parser.deref(schemaOrRef);
    this.schema = parser.mergeAllOf(this.rawSchema, this.pointer, isChild);
    this.init(parser, isChild);

    parser.exitRef(schemaOrRef);

    for (const parent$ref of this.schema.parentRefs || []) {
      // exit all the refs visited during allOf traverse
      parser.exitRef({ $ref: parent$ref });
    }

    if (options.showExtensions) {
      this.extensions = extractExtensions(this.schema, options.showExtensions);
    }
  }

  /**
   * Set specified alternative schema as active
   * @param idx oneOf index
   */
  @action
  activateOneOf(idx: number) {
    this.activeOneOf = idx;
  }

  init(parser: OpenAPIParser, isChild: boolean) {
    const schema = this.schema;
    this.isCircular = schema['x-circular-ref'];

    this.title =
      schema.title || (isNamedDefinition(this.pointer) && JsonPointer.baseName(this.pointer)) || '';
    this.description = schema.description || '';
    this.type = schema.type || detectType(schema);
    this.format = schema.format;
    this.nullable = !!schema.nullable;
    this.enum = schema.enum || [];
    this.example = schema.example;
    this.deprecated = !!schema.deprecated;
    this.pattern = schema.pattern;
    this.externalDocs = schema.externalDocs;

    this.constraints = humanizeConstraints(schema);
    this.displayType = this.type;
    this.displayFormat = this.format;
    this.isPrimitive = isPrimitiveType(schema, this.type);
    this.default = schema.default;
    this.readOnly = !!schema.readOnly;
    this.writeOnly = !!schema.writeOnly;

    if (this.isCircular) {
      return;
    }

    if (!isChild && getDiscriminator(schema) !== undefined) {
      this.initDiscriminator(schema, parser);
      return;
    }

    if (schema.oneOf !== undefined) {
      this.initOneOf(schema.oneOf, parser);
      this.oneOfType = 'One of';
      if (schema.anyOf !== undefined) {
        console.warn(
          `oneOf and anyOf are not supported on the same level. Skipping anyOf at ${this.pointer}`,
        );
      }
      return;
    }

    if (schema.anyOf !== undefined) {
      this.initOneOf(schema.anyOf, parser);
      this.oneOfType = 'Any of';
      return;
    }

    if (this.type === 'object') {
      this.fields = buildFields(parser, schema, this.pointer, this.options);
    } else if (this.type === 'array' && schema.items) {
      this.items = new SchemaModel(parser, schema.items, this.pointer + '/items', this.options);
      this.displayType = this.items.displayType;
      this.displayFormat = this.items.format;
      this.typePrefix = this.items.typePrefix + 'Array of ';
      this.title = this.title || this.items.title;
      this.isPrimitive = this.items.isPrimitive;
      if (this.example === undefined && this.items.example !== undefined) {
        this.example = [this.items.example];
      }
      if (this.items.isPrimitive) {
        this.enum = this.items.enum;
      }
    }
  }

  private initOneOf(oneOf: OpenAPISchema[], parser: OpenAPIParser) {
    this.oneOf = oneOf!.map(
      (variant, idx) =>
        new SchemaModel(
          parser,
          // merge base schema into each of oneOf's subschemas
          {
            // variant may already have allOf so merge it to not get overwritten
            ...parser.mergeAllOf(variant, this.pointer + '/oneOf/' + idx),
            allOf: [{ ...this.schema, oneOf: undefined, anyOf: undefined }],
          } as OpenAPISchema,
          this.pointer + '/oneOf/' + idx,
          this.options,
        ),
    );
    this.displayType = this.oneOf
      .map(schema => {
        let name =
          schema.typePrefix +
          (schema.title ? `${schema.title} (${schema.displayType})` : schema.displayType);
        if (name.indexOf(' or ') > -1) {
          name = `(${name})`;
        }
        return name;
      })
      .join(' or ');
  }

  private initDiscriminator(
    schema: OpenAPISchema & {
      parentRefs?: string[];
    },
    parser: OpenAPIParser,
  ) {
    const discriminator = getDiscriminator(schema)!;
    this.discriminatorProp = discriminator.propertyName;
    const derived = parser.findDerived([...(schema.parentRefs || []), this.pointer]);

    if (schema.oneOf) {
      for (const variant of schema.oneOf) {
        if (variant.$ref === undefined) {
          continue;
        }
        const name = JsonPointer.dirName(variant.$ref);
        derived[variant.$ref] = name;
      }
    }

    const mapping = discriminator.mapping || {};
    for (const name in mapping) {
      derived[mapping[name]] = name;
    }

    const refs = Object.keys(derived);
    this.oneOf = refs.map(ref => {
      const innerSchema = new SchemaModel(parser, parser.byRef(ref)!, ref, this.options, true);
      innerSchema.title = derived[ref];
      return innerSchema;
    });
  }
}

function buildFields(
  parser: OpenAPIParser,
  schema: OpenAPISchema,
  $ref: string,
  options: RedocNormalizedOptions,
): FieldModel[] {
  const props = schema.properties || {};
  const additionalProps = schema.additionalProperties;
  const defaults = schema.default || {};
  const fields = Object.keys(props || []).map(fieldName => {
    let field = props[fieldName];

    if (!field) {
      console.warn(
        `Field "${fieldName}" is invalid, skipping.\n Field must be an object but got ${typeof field} at "${$ref}"`,
      );
      field = {};
    }

    const required =
      schema.required === undefined ? false : schema.required.indexOf(fieldName) > -1;

    return new FieldModel(
      parser,
      {
        name: fieldName,
        required,
        schema: {
          ...field,
          default: field.default === undefined ? defaults[fieldName] : field.default,
        },
      },
      $ref + '/properties/' + fieldName,
      options,
    );
  });

  if (options.sortPropsAlphabetically) {
    sortByField(fields, 'name');
  }
  if (options.requiredPropsFirst) {
    // if not sort alphabetically sort in the order from required keyword
    sortByRequired(fields, !options.sortPropsAlphabetically ? schema.required : undefined);
  }

  if (typeof additionalProps === 'object' || additionalProps === true) {
    fields.push(
      new FieldModel(
        parser,
        {
          name: 'property name *',
          required: false,
          schema: additionalProps === true ? {} : additionalProps,
          kind: 'additionalProperties',
        },
        $ref + '/additionalProperties',
        options,
      ),
    );
  }

  return fields;
}

function getDiscriminator(schema: OpenAPISchema): OpenAPISchema['discriminator'] {
  return schema.discriminator || schema['x-discriminator'];
}
