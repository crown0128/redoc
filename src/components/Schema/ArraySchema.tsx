import * as React from 'react';

import { Schema, SchemaProps } from './Schema';

import { ArrayClosingLabel, ArrayOpenningLabel } from '../../common-elements';
import styled from '../../styled-components';
import {humanizeConstraints} from "../../utils";

const PaddedSchema = styled.div`
  padding-left: ${({ theme }) => theme.spacing.unit * 2}px;
`;

export class ArraySchema extends React.PureComponent<SchemaProps> {
  render() {
    const itemsSchema = this.props.schema.items!;
    const itemConstraintSchema = (
      min: number | undefined = undefined,
      max: number | undefined = undefined,
    ) => ({ type: 'array', minItems: min, maxItems: max });

    const minMaxItems = humanizeConstraints(itemConstraintSchema(itemsSchema.schema.minItems, itemsSchema.schema.maxItems));

    return (
      <div>
        <ArrayOpenningLabel> Array ({minMaxItems})</ArrayOpenningLabel>
        <PaddedSchema>
          <Schema {...this.props} schema={itemsSchema} />
        </PaddedSchema>
        <ArrayClosingLabel />
      </div>
    );
  }
}
