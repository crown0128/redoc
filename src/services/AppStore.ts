import { observe } from 'mobx';

import { OpenAPISpec } from '../types';
import { loadAndBundleSpec } from '../utils/loadAndBundleSpec';
import { MenuStore } from './MenuStore';
import { SpecStore } from './models';
import { RedocNormalizedOptions, RedocRawOptions } from './RedocNormalizedOptions';
import { ScrollService } from './ScrollService';
import { SearchStore } from './SearchStore';
import { MarkerService } from './MarkerService';

interface StoreData {
  menu: {
    activeItemIdx: number;
  };
  spec: {
    url: string;
    data: any;
  };
  options: RedocRawOptions;
}

export async function createStore(
  spec: object,
  specUrl: string | undefined,
  options: RedocRawOptions = {},
) {
  const resolvedSpec = await loadAndBundleSpec(spec || specUrl);
  return new AppStore(resolvedSpec, specUrl, options);
}

export class AppStore {
  /**
   * deserialize store
   * **SUPER HACKY AND NOT OPTIMAL IMPLEMENTATION**
   */
  // TODO:
  static fromJS(state: StoreData): AppStore {
    const inst = new AppStore(state.spec.data, state.spec.url, state.options);
    inst.menu.activeItemIdx = state.menu.activeItemIdx || 0;
    inst.menu.activate(inst.menu.flatItems[inst.menu.activeItemIdx]);
    return inst;
  }

  menu: MenuStore;
  spec: SpecStore;
  rawOptions: RedocRawOptions;
  options: RedocNormalizedOptions;
  search: SearchStore;
  marker = new MarkerService();

  private scroll: ScrollService;
  private disposer;

  constructor(spec: OpenAPISpec, specUrl?: string, options: RedocRawOptions = {}) {
    this.rawOptions = options;
    this.options = new RedocNormalizedOptions(options);
    this.scroll = new ScrollService(this.options);
    this.spec = new SpecStore(spec, specUrl, this.options);
    this.menu = new MenuStore(this.spec, this.scroll);

    this.search = new SearchStore(this.spec);

    this.disposer = observe(this.menu, 'activeItemIdx', change => {
      this.updateMarkOnMenu(change.newValue as number);
    });
  }

  updateMarkOnMenu(idx: number) {
    console.log('update marker');
    const start = Math.max(0, idx);
    const end = Math.min(this.menu.flatItems.length, start + 5);

    const elements: Element[] = [];
    for (let i = start; i < end; i++) {
      let elem = this.menu.getElementAt(i);
      if (!elem) continue;
      if (this.menu.flatItems[i].type === 'section') {
        elem = elem.parentElement!.parentElement;
      }
      if (elem) elements.push(elem);
    }

    this.marker.addOnly(elements);
    this.marker.mark();
  }

  dispose() {
    this.scroll.dispose();
    this.menu.dispose();
    this.disposer();
  }

  /**
   * serializes store
   * **SUPER HACKY AND NOT OPTIMAL IMPLEMENTATION**
   */
  // TODO:
  toJS(): StoreData {
    return {
      menu: {
        activeItemIdx: this.menu.activeItemIdx,
      },
      spec: {
        url: this.spec.parser.specUrl,
        data: this.spec.parser.spec,
      },
      options: this.rawOptions,
    };
  }
}
