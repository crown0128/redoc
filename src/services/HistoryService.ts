import { bind, debounce } from 'decko';
import { EventEmitter } from 'eventemitter3';
import { IS_BROWSER } from '../utils/';

const EVENT = 'hashchange';

function isSameHash(a: string, b: string): boolean {
  return a === b || '#' + a === b || a === '#' + b;
}

export class IntHistoryService {
  private causedHashChange: boolean = false;
  private _emiter;

  constructor() {
    this._emiter = new EventEmitter();
    this.bind();
  }

  get hash(): string {
    return IS_BROWSER ? window.location.hash : '';
  }

  subscribe(cb): () => void {
    const emmiter = this._emiter.addListener(EVENT, cb);
    return () => emmiter.removeListener(EVENT, cb);
  }

  emit = () => {
    if (this.causedHashChange) {
      this.causedHashChange = false;
      return;
    }
    this._emiter.emit(EVENT, this.hash);
  };

  bind() {
    if (IS_BROWSER) {
      window.addEventListener('hashchange', this.emit, false);
    }
  }

  dispose() {
    if (IS_BROWSER) {
      window.removeEventListener('hashchange', this.emit);
    }
    this.causedHashChange = false;
  }

  @bind
  @debounce
  update(hash: string | null, rewriteHistory: boolean = false) {
    if (hash == null || isSameHash(hash, this.hash)) {
      return;
    }
    if (rewriteHistory) {
      if (IS_BROWSER) {
        window.history.replaceState(null, '', window.location.href.split('#')[0] + '#' + hash);
      }
      return;
    }
    this.causedHashChange = true;
    if (IS_BROWSER) {
      window.location.hash = hash;
    }
  }
}

export const HistoryService = new IntHistoryService();

if (module.hot) {
  module.hot.dispose(() => {
    HistoryService.dispose();
  });
}
