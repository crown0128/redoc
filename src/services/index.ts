export * from './AppStore';
export * from './OpenAPIParser';
export * from './MarkdownRenderer';
export * from './MenuStore';
export * from './ScrollService';
export * from './SpecStore';
export * from './ClipboardService';
export * from './HistoryService';
export * from './models';
