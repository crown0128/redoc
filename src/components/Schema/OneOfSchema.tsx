import * as React from 'react';
import { observer } from 'mobx-react';

import { OneOfButton, OneOfLabel, OneOfList } from '../../common-elements/schema';
import { SchemaProps, Schema } from './Schema';

@observer
export class OneOfSchema extends React.Component<SchemaProps> {
  render() {
    const { schema: { oneOf }, schema } = this.props;

    if (oneOf === undefined) {
      return null;
    }
    return (
      <div>
        <OneOfLabel> {schema.oneOfType} </OneOfLabel>
        <OneOfList>
          {oneOf.map((subSchema, idx) => (
            <OneOfButton
              key={subSchema._$ref}
              active={idx === schema.activeOneOf}
              onClick={() => schema.activateOneOf(idx)}
            >
              {subSchema.title || subSchema.displayType}
            </OneOfButton>
          ))}
        </OneOfList>
        <Schema schema={oneOf[schema.activeOneOf]} />
      </div>
    );
  }
}
