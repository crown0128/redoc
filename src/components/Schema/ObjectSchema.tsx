import { observer } from 'mobx-react';
import * as React from 'react';

import { FieldModel, SchemaModel } from '../../services/models';

import {
  InnerPropertiesWrap,
  PropertiesTable,
  PropertiesTableCaption,
  PropertyCellWithInner,
} from '../../common-elements/fields-layout';
import { Field } from '../Fields/Field';
import { DiscriminatorDropdown } from './DiscriminatorDropdown';
import { Schema, SchemaProps } from './Schema';

import { mapWithLast } from '../../utils';

export interface ObjectSchemaProps extends SchemaProps {
  discriminator?: {
    fieldName: string;
    parentSchema: SchemaModel;
  };
}

@observer
export class ObjectSchema extends React.Component<ObjectSchemaProps> {
  get parentSchema() {
    return this.props.discriminator!.parentSchema;
  }

  renderField(field: FieldModel, isLast: boolean, isDiscriminator: boolean = false) {
    const withSubSchema = !field.schema.isPrimitive && !field.schema.isCircular;
    return [
      <Field
        key={field.name}
        isLast={isLast}
        field={field}
        onClick={(withSubSchema && (() => field.toggle())) || undefined}
        renderDiscriminatorSwitch={
          (isDiscriminator &&
            (() => (
              <DiscriminatorDropdown parent={this.parentSchema} enumValues={field.schema.enum} />
            ))) ||
          undefined
        }
        className={field.expanded ? 'expanded' : undefined}
        showExamples={false}
      />,
      field.expanded &&
        withSubSchema && (
          <tr key={field.name + 'inner'}>
            <PropertyCellWithInner colSpan={2}>
              <InnerPropertiesWrap>
                <Schema
                  schema={field.schema}
                  skipReadOnly={this.props.skipReadOnly}
                  skipWriteOnly={this.props.skipWriteOnly}
                  showTitle={this.props.showTitle}
                />
              </InnerPropertiesWrap>
            </PropertyCellWithInner>
          </tr>
        ),
    ];
  }

  render() {
    const { schema: { fields = [] }, showTitle, discriminator } = this.props;

    const needFilter = this.props.skipReadOnly || this.props.skipWriteOnly;

    const filteredFields = needFilter
      ? fields.filter(item => {
          return (
            (this.props.skipReadOnly && !item.schema.readOnly) ||
            (this.props.skipWriteOnly && !item.schema.writeOnly)
          );
        })
      : fields;

    return (
      <PropertiesTable>
        {showTitle && <PropertiesTableCaption>{this.props.schema.title}</PropertiesTableCaption>}
        <tbody>
          {mapWithLast(filteredFields, (field, isLast) =>
            this.renderField(
              field,
              isLast,
              discriminator && discriminator.fieldName === field.name,
            ),
          )}
        </tbody>
      </PropertiesTable>
    );
  }
}
