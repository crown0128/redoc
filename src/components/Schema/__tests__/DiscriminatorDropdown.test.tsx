import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { OpenAPIParser, SchemaModel } from '../../../services';
import { Schema } from '../Schema';
import { ObjectSchema } from '../ObjectSchema';
import * as simpleDiscriminatorFixture from './fixtures/simple-discriminator.json';

describe('Components', () => {
  describe('SchemaView', () => {
    describe('discriminator', () => {
      it('should correctly render SchemaView', () => {
        const parser = new OpenAPIParser(simpleDiscriminatorFixture);

        const schema = new SchemaModel(
          parser,
          { $ref: '#/components/schemas/Pet' },
          '#/components/schemas/Pet',
        );
        const schemaView = shallow(<Schema schema={schema} />);
        expect(toJson(schemaView)).toMatchSnapshot();
      });

      it('should correctly render discriminator dropdown', () => {
        const parser = new OpenAPIParser(simpleDiscriminatorFixture);

        const schema = new SchemaModel(
          parser,
          { $ref: '#/components/schemas/Pet' },
          '#/components/schemas/Pet',
        );
        const schemaView = shallow(
          <ObjectSchema
            schema={schema.oneOf![0]}
            discriminator={{
              fieldName: schema.discriminatorProp,
              parentSchema: schema,
            }}
          />,
        );
        expect(toJson(schemaView)).toMatchSnapshot();
      });
    });
  });
});
