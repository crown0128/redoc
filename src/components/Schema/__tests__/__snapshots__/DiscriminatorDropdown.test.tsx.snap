// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components SchemaView discriminator should correctly render SchemaView 1`] = `
<ObjectSchema
  discriminator={
    Object {
      "fieldName": "type",
      "parentSchema": SchemaModel {
        "_$ref": "#/components/schemas/Pet",
        "activeOneOf": 0,
        "constraints": Array [],
        "default": undefined,
        "deprecated": false,
        "description": "",
        "discriminatorProp": "type",
        "displayType": "Pet (object)",
        "enum": Array [],
        "example": undefined,
        "format": undefined,
        "isCircular": undefined,
        "isPrimitive": false,
        "nullable": false,
        "oneOf": Array [
          SchemaModel {
            "_$ref": "#/components/schemas/Dog",
            "activeOneOf": 0,
            "constraints": Array [],
            "default": undefined,
            "deprecated": false,
            "description": "",
            "displayType": "Dog (object)",
            "enum": Array [],
            "example": undefined,
            "fields": Array [
              FieldModel {
                "deprecated": false,
                "description": "",
                "example": undefined,
                "expanded": false,
                "in": undefined,
                "name": "packSize",
                "required": false,
                "schema": SchemaModel {
                  "_$ref": "#/components/schemas/Dog/properties/packSize/schema",
                  "activeOneOf": 0,
                  "constraints": Array [],
                  "default": undefined,
                  "deprecated": false,
                  "description": "",
                  "displayType": "number",
                  "enum": Array [],
                  "example": undefined,
                  "format": undefined,
                  "isCircular": undefined,
                  "isPrimitive": true,
                  "nullable": false,
                  "pattern": undefined,
                  "rawSchema": Object {
                    "default": undefined,
                    "type": "number",
                  },
                  "schema": Object {
                    "default": undefined,
                    "type": "number",
                  },
                  "title": "",
                  "type": "number",
                  "typePrefix": "",
                },
              },
              FieldModel {
                "deprecated": false,
                "description": "",
                "example": undefined,
                "expanded": false,
                "in": undefined,
                "name": "type",
                "required": true,
                "schema": SchemaModel {
                  "_$ref": "#/components/schemas/Dog/properties/type/schema",
                  "activeOneOf": 0,
                  "constraints": Array [],
                  "default": undefined,
                  "deprecated": false,
                  "description": "",
                  "displayType": "string",
                  "enum": Array [],
                  "example": undefined,
                  "format": undefined,
                  "isCircular": undefined,
                  "isPrimitive": true,
                  "nullable": false,
                  "pattern": undefined,
                  "rawSchema": Object {
                    "default": undefined,
                    "type": "string",
                  },
                  "schema": Object {
                    "default": undefined,
                    "type": "string",
                  },
                  "title": "",
                  "type": "string",
                  "typePrefix": "",
                },
              },
            ],
            "format": undefined,
            "isCircular": undefined,
            "isPrimitive": false,
            "nullable": false,
            "pattern": undefined,
            "rawSchema": Object {
              "allOf": Array [
                Object {
                  "$ref": "#/components/schemas/Pet",
                },
              ],
              "properties": Object {
                "packSize": Object {
                  "type": "number",
                },
              },
              "type": "object",
            },
            "schema": Object {
              "allOf": undefined,
              "discriminator": Object {
                "propertyName": "type",
              },
              "namedParents": Array [
                "#/components/schemas/Pet",
              ],
              "properties": Object {
                "packSize": Object {
                  "type": "number",
                },
                "type": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "type",
              ],
              "title": "Dog",
              "type": "object",
            },
            "title": "Dog",
            "type": "object",
            "typePrefix": "",
          },
          SchemaModel {
            "_$ref": "#/components/schemas/Cat",
            "activeOneOf": 0,
            "constraints": Array [],
            "default": undefined,
            "deprecated": false,
            "description": "",
            "displayType": "Cat (object)",
            "enum": Array [],
            "example": undefined,
            "fields": Array [
              FieldModel {
                "deprecated": false,
                "description": "",
                "example": undefined,
                "expanded": false,
                "in": undefined,
                "name": "type",
                "required": true,
                "schema": SchemaModel {
                  "_$ref": "#/components/schemas/Cat/properties/type/schema",
                  "activeOneOf": 0,
                  "constraints": Array [],
                  "default": undefined,
                  "deprecated": false,
                  "description": "",
                  "displayType": "string",
                  "enum": Array [],
                  "example": undefined,
                  "format": undefined,
                  "isCircular": undefined,
                  "isPrimitive": true,
                  "nullable": false,
                  "pattern": undefined,
                  "rawSchema": Object {
                    "default": undefined,
                    "type": "string",
                  },
                  "schema": Object {
                    "default": undefined,
                    "type": "string",
                  },
                  "title": "",
                  "type": "string",
                  "typePrefix": "",
                },
              },
              FieldModel {
                "deprecated": false,
                "description": "",
                "example": undefined,
                "expanded": false,
                "in": undefined,
                "name": "packSize",
                "required": false,
                "schema": SchemaModel {
                  "_$ref": "#/components/schemas/Cat/properties/packSize/schema",
                  "activeOneOf": 0,
                  "constraints": Array [],
                  "default": undefined,
                  "deprecated": false,
                  "description": "",
                  "displayType": "number",
                  "enum": Array [],
                  "example": undefined,
                  "format": undefined,
                  "isCircular": undefined,
                  "isPrimitive": true,
                  "nullable": false,
                  "pattern": undefined,
                  "rawSchema": Object {
                    "default": undefined,
                    "type": "number",
                  },
                  "schema": Object {
                    "default": undefined,
                    "type": "number",
                  },
                  "title": "",
                  "type": "number",
                  "typePrefix": "",
                },
              },
            ],
            "format": undefined,
            "isCircular": undefined,
            "isPrimitive": true,
            "nullable": false,
            "pattern": undefined,
            "rawSchema": Object {
              "allOf": Array [
                Object {
                  "$ref": "#/components/schemas/Pet",
                },
                Object {
                  "properties": Object {
                    "packSize": Object {
                      "type": "number",
                    },
                  },
                },
              ],
              "type": "object",
            },
            "schema": Object {
              "allOf": undefined,
              "discriminator": Object {
                "propertyName": "type",
              },
              "namedParents": Array [
                "#/components/schemas/Pet",
              ],
              "properties": Object {
                "packSize": Object {
                  "type": "number",
                },
                "type": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "type",
              ],
              "title": "Cat",
              "type": undefined,
            },
            "title": "Cat",
            "type": "object",
            "typePrefix": "",
          },
        ],
        "pattern": undefined,
        "rawSchema": Object {
          "discriminator": Object {
            "propertyName": "type",
          },
          "properties": Object {
            "type": Object {
              "type": "string",
            },
          },
          "required": Array [
            "type",
          ],
          "type": "object",
        },
        "schema": Object {
          "discriminator": Object {
            "propertyName": "type",
          },
          "properties": Object {
            "type": Object {
              "type": "string",
            },
          },
          "required": Array [
            "type",
          ],
          "type": "object",
        },
        "title": "Pet",
        "type": "object",
        "typePrefix": "",
      },
    }
  }
  schema={
    SchemaModel {
      "_$ref": "#/components/schemas/Dog",
      "activeOneOf": 0,
      "constraints": Array [],
      "default": undefined,
      "deprecated": false,
      "description": "",
      "displayType": "Dog (object)",
      "enum": Array [],
      "example": undefined,
      "fields": Array [
        FieldModel {
          "deprecated": false,
          "description": "",
          "example": undefined,
          "expanded": false,
          "in": undefined,
          "name": "packSize",
          "required": false,
          "schema": SchemaModel {
            "_$ref": "#/components/schemas/Dog/properties/packSize/schema",
            "activeOneOf": 0,
            "constraints": Array [],
            "default": undefined,
            "deprecated": false,
            "description": "",
            "displayType": "number",
            "enum": Array [],
            "example": undefined,
            "format": undefined,
            "isCircular": undefined,
            "isPrimitive": true,
            "nullable": false,
            "pattern": undefined,
            "rawSchema": Object {
              "default": undefined,
              "type": "number",
            },
            "schema": Object {
              "default": undefined,
              "type": "number",
            },
            "title": "",
            "type": "number",
            "typePrefix": "",
          },
        },
        FieldModel {
          "deprecated": false,
          "description": "",
          "example": undefined,
          "expanded": false,
          "in": undefined,
          "name": "type",
          "required": true,
          "schema": SchemaModel {
            "_$ref": "#/components/schemas/Dog/properties/type/schema",
            "activeOneOf": 0,
            "constraints": Array [],
            "default": undefined,
            "deprecated": false,
            "description": "",
            "displayType": "string",
            "enum": Array [],
            "example": undefined,
            "format": undefined,
            "isCircular": undefined,
            "isPrimitive": true,
            "nullable": false,
            "pattern": undefined,
            "rawSchema": Object {
              "default": undefined,
              "type": "string",
            },
            "schema": Object {
              "default": undefined,
              "type": "string",
            },
            "title": "",
            "type": "string",
            "typePrefix": "",
          },
        },
      ],
      "format": undefined,
      "isCircular": undefined,
      "isPrimitive": false,
      "nullable": false,
      "pattern": undefined,
      "rawSchema": Object {
        "allOf": Array [
          Object {
            "$ref": "#/components/schemas/Pet",
          },
        ],
        "properties": Object {
          "packSize": Object {
            "type": "number",
          },
        },
        "type": "object",
      },
      "schema": Object {
        "allOf": undefined,
        "discriminator": Object {
          "propertyName": "type",
        },
        "namedParents": Array [
          "#/components/schemas/Pet",
        ],
        "properties": Object {
          "packSize": Object {
            "type": "number",
          },
          "type": Object {
            "type": "string",
          },
        },
        "required": Array [
          "type",
        ],
        "title": "Dog",
        "type": "object",
      },
      "title": "Dog",
      "type": "object",
      "typePrefix": "",
    }
  }
/>
`;

exports[`Components SchemaView discriminator should correctly render discriminator dropdown 1`] = `
<styled.table>
  <tbody>
    <Field
      field={
        FieldModel {
          "deprecated": false,
          "description": "",
          "example": undefined,
          "expanded": false,
          "in": undefined,
          "name": "packSize",
          "required": false,
          "schema": SchemaModel {
            "_$ref": "#/components/schemas/Dog/properties/packSize/schema",
            "activeOneOf": 0,
            "constraints": Array [],
            "default": undefined,
            "deprecated": false,
            "description": "",
            "displayType": "number",
            "enum": Array [],
            "example": undefined,
            "format": undefined,
            "isCircular": undefined,
            "isPrimitive": true,
            "nullable": false,
            "pattern": undefined,
            "rawSchema": Object {
              "default": undefined,
              "type": "number",
            },
            "schema": Object {
              "default": undefined,
              "type": "number",
            },
            "title": "",
            "type": "number",
            "typePrefix": "",
          },
        }
      }
      isLast={false}
      key="packSize"
      showExamples={false}
    />
    <Field
      field={
        FieldModel {
          "deprecated": false,
          "description": "",
          "example": undefined,
          "expanded": false,
          "in": undefined,
          "name": "type",
          "required": true,
          "schema": SchemaModel {
            "_$ref": "#/components/schemas/Dog/properties/type/schema",
            "activeOneOf": 0,
            "constraints": Array [],
            "default": undefined,
            "deprecated": false,
            "description": "",
            "displayType": "string",
            "enum": Array [],
            "example": undefined,
            "format": undefined,
            "isCircular": undefined,
            "isPrimitive": true,
            "nullable": false,
            "pattern": undefined,
            "rawSchema": Object {
              "default": undefined,
              "type": "string",
            },
            "schema": Object {
              "default": undefined,
              "type": "string",
            },
            "title": "",
            "type": "string",
            "typePrefix": "",
          },
        }
      }
      isLast={true}
      key="type"
      renderDiscriminatorSwitch={[Function]}
      showExamples={false}
    />
  </tbody>
</styled.table>
`;
